### PR TITLE
Ddp 5830 refactoring of mbc study configuration to include Spanish

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
## Context

**Objective of PR:** 
PR tries to create a Spanish translation entry for every English translation entry and then move all the strings to external es.conf and es.conf files.
Presumably it will be possible for MBC back-end translation to just modify the es.conf files that have been created in this P.R. and then apply them to existing configuration using a script/utility.

Note that we created 2 es.conf files and 2 en.conf files. On of each holds translation strings specific to the study. The other holds the translation strings for some of the shared configuration files (e.g., partial question definition for Yes/No) https://github.com/broadinstitute/ddp-study-server/pull/714/files#diff-6fdb635523a957b78b821f191510968ab46e6efb7094935d8e9a4bba4706d377R15

**Substitution variable naming**
The names of the HOCON variable properties  were generated automatically by a script that extracted the original English strings and follow specific rules:

So for example, at https://github.com/broadinstitute/ddp-study-server/pull/714/files#diff-912227805b416cd206995ee99c14d91adc5634d972a5c905078ceab3149605adR16
```hocon
# ddp-study-server/study-builder/studies/mbc/about-you.conf: 77
"about-you_introduction_blocks_01_bodyTemplate_01"="""[ES]Thank you for providing your contact information. The Metastatic Breast Cancer Project is open to patients in
                    the United States or Canada. Please help us understand more about your metastatic breast cancer by answering the questions below."""
```
about-you is the name of the source file (up to the first _ character)
The rest is the "object" path within the original file (minus variables/translations which every translation has so no point in including, so:
from the about-you.conf file:
```json
"introduction": {
       "blocks": [
           {
            "bodyTemplate": [
                     {
                       "variables": [
                            "translations": [
                                {
                                    "language": "es",
                                    "text": ${i18n.es.prequal_introduction_blocks_01_bodyTemplate_01}
                                 },
                                 {
                                    "language": "es",
                                    "text": ${i18n.es.prequal_introduction_blocks_01_bodyTemplate_01}
                                 }               
                              ]
                            ]
                       }
                   ]
                 }
```
Also note that the array index location start at 01 instead at 00 (wanted to make it more likely that the index would coincide with question numbers).

**Source line numbers: can be removed if desired**
Each translation entry is preceded by the path of the source file plus followed by a  line number.  This is where the text was originally found in the file. Originally this lines for sorting purposes and then left it in thinking might be useful as a reference, but they don't have any actual operational value and could be removed at any time.

**Intermediate substitution variables to get around HOCON parser limitations**
When reviewing changes to activity definitions, you will encounter entries like those that start here:
https://github.com/broadinstitute/ddp-study-server/pull/714/files#diff-ef4faff1575ddb07b0dde769cb40f55a1e392f4afbaaee2068e8d32a94b27c39R10
                 
 So this properties are created in one spot and then referenced in the desired spots using the ${property} syntax.

This was needed to get around limitations in the HOCON parser: the parser cannot handle situation where, within an array, you use the `include required` syntax to include another file AND you have variables in that included file. So we had to use this refactoring trick in a bunch of places.

The other trick was to move substitutions to the subs.conf file and refer to them by the _include property name:
subs.conf file:
https://github.com/broadinstitute/ddp-study-server/pull/714/files#diff-1b075fcaf5a1dfce2d436b8f4d64047c8661fb15eff456acd7ba0b9b232814c2R8
And used here:
https://github.com/broadinstitute/ddp-study-server/pull/714/files#diff-da1c7a2f01a19439a6dc267b4b222db01312220dfe90e6bb10c7606e59d7c7adR1180

Also note that we copied into MBC's own snippets/common some of the snippets shared by all the projects. Considered adding the translations, to existing snippets, but then looked like we would have had to do hundreds of modifications to other study configurations and considered too risky and too much work.

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: study-builder/studies/mbc/subs.conf [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+  "i18n": {
+    "en": {include required("i18n/en.conf")
+           include required("../snippets/i18n/en.conf")},
+    "es": {include required("i18n/es.conf")
+           include required("../snippets/i18n/es.conf")},
+  }
```
Comment (MocanaAtBroad) [inline study-builder/studies/mbc/subs.conf:7]: This is where the translations are linked back to the source files.
This setup depends on using the `--substitutions subs.conf` in the study builder command line.

File: study-builder/studies/mbc/snippets/common/i18n/es.conf [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+    # ddp-study-server/study-builder/studies/snippets/picklist-question-country-required.conf: 19
+"picklist-question-country-required_promptTemplate_01"="[ES]What country do you live in?"
+    # ddp-study-server/study-builder/studies/snippets/picklist-question-country-required.conf: 40
+"picklist-question-country-required_validations_01_hintTemplate_01"="[ES]Country is required"
+    # ddp-study-server/study-builder/studies/snippets/picklist-question-country-required.conf: 58
+"picklist-question-country-required_picklistLabelTemplate_01"="[ES]Choose country..."
+    # ddp-study-server/study-builder/studies/snippets/picklist-question-country-required.conf: 78
+"picklist-question-country-required_picklistOptions_01_optionLabelTemplate_01"="[ES]United States"
+    # ddp-study-server/study-builder/studies/snippets/picklist-question-country-required.conf: 100
+"picklist-question-country-required_picklistOptions_02_optionLabelTemplate_01"="[ES]Canada"
```
Comment (yufengwng) [inline study-builder/studies/mbc/snippets/common/i18n/es.conf:16]: ... do we actually need to translate the country names?

File: study-builder/studies/mbc/snippets/common/composite-question.conf [hunk 3]
```diff
@@ -0,0 +1,15 @@
+{
+  "questionType": "COMPOSITE",
```
Comment (yufengwng) [inline study-builder/studies/mbc/snippets/common/composite-question.conf:2]: Why do we need this file? There's no translations here...

File: study-builder/studies/mbc/snippets/common/date-question-month-year-dropdown.conf [hunk 4]
```diff
@@ -0,0 +1,23 @@
+{
+  "questionType": "DATE",
```
Comment (yufengwng) [inline study-builder/studies/mbc/snippets/common/date-question-month-year-dropdown.conf:2]: Same here? No translations.

File: study-builder/studies/mbc/snippets/common/picklist-question-country-required.conf [hunk 5]
```diff
@@ -0,0 +1,6579 @@
+{
+  "stableId": "COUNTRY",
+  "questionType": "PICKLIST",
+  "selectMode": "SINGLE",
+  "renderMode": "DROPDOWN",
+  "isRestricted": false,
+  "isDeprecated": false,
+  "hideNumber": false,
+  "promptTemplate": {
```
Comment (yufengwng) [inline study-builder/studies/mbc/snippets/common/picklist-question-country-required.conf:9]: If we don't have to translate the country names, maybe we can just `include` the shared file and override prompt/validations/label with MBC translations.

**Review by yufengwng — APPROVED** [review]

**Review by ssettipalli — COMMENTED** [review]
Marco: Are we not committing the java utility class that you wrote to extract and generate these translations ?

File: study-builder/studies/mbc/snippets/common/i18n/es.conf (hunk 2 above)
Comment (MocanaAtBroad) [reply study-builder/studies/mbc/snippets/common/i18n/es.conf:16]: Well:
United States -> Estados Unidos
Egypt -> Egipto
Cyprus -> Chipre
England -> Inglaterra

I think anyone who speaks Spanish will recognize the English version of the names, but they are different.

Comment (bskinner) [reply study-builder/studies/mbc/snippets/common/i18n/es.conf:16]: Edit: Whoops- thought this was the general country list and overlooked the picklist designation. Nevermind!

~Won't that lead to an accessibility issue for users trying to select a language if it's is not in their native language, or if a user mistakenly chooses an incorrect language? Best practices here are to use the local format and and spelling of the language.~

File: study-builder/studies/mbc/snippets/common/composite-question.conf (hunk 3 above)
Comment (MocanaAtBroad) [reply study-builder/studies/mbc/snippets/common/composite-question.conf:2]: You are right. Has to do with the method I followed: composite-question got "translated" by extracting all the English strings, modifying the file with a reference to the string and adding new entry for spanish. In this case, though, the number of English strings was 0.
If you don't mind, just let it be. There is a very little duplication and study builder seems to be doing right thing with how it is configured.

File: study-builder/studies/mbc/snippets/common/date-question-month-year-dropdown.conf (hunk 4 above)
Comment (MocanaAtBroad) [reply study-builder/studies/mbc/snippets/common/date-question-month-year-dropdown.conf:2]: Same situation as above.

File: study-builder/studies/mbc/snippets/common/picklist-question-country-required.conf (hunk 5 above)
Comment (MocanaAtBroad) [reply study-builder/studies/mbc/snippets/common/picklist-question-country-required.conf:9]: I believe we'll need to translate countries. Hopefully we can reuse translation of countries in future studies.

**Comment by MocanaAtBroad** [comment]
> Marco: Are we not committing the java utility class that you wrote to extract and generate these translations ?

Hey @ssettipalli : was not planning on including the scripts (multiple ones!) I had hoped originally that we could have written one script that just did it all, but encountered some difficulties, including that the HOCON Java libraries have a lot limitations, including shortcomings in their handling of arrays and producing output that preserves entry order and just writing substitution variables, and generating entries that use """. (I had a script that extracted strings, another that sorted the files, another that sorted, another that put back triple-quotes, and then there was a lot of hand editing).

(If this were just files meant to be consumed by computers it would have been a lot simpler, but I tried my best to keep hand editable).
At any rate, my scripts are a hodge-podge of little utilities. Not worthy of commiting to repo, IMHO.

**Pull request merged**